internal/users/idgenerator.go

```diff
@@ -0,0 +1,278 @@
+		getUsedIDs:    g.getUsedGIDs,
+		isAvailableID: g.isGIDAvailable,
+	})
+}
+
+// This is an utility struct to allow code sharing simplifying the arguments passing.
+type generateID struct {
+	idType        string
+	minID, maxID  uint32
+	isAvailableID func(*localentries.UserDBLocked, uint32) (bool, error)
+	getUsedIDs    func(*localentries.UserDBLocked, IDOwner) ([]uint32, error)
+}
+
+func (g *IDGenerator) generateID(lockedEntries *localentries.UserDBLocked, owner IDOwner, args generateID) (id uint32, cleanup func(), err error) {
+	if args.minID > args.maxID {
+		return 0, nil, errors.New("minID must be less than or equal to maxID")
+	}
+
+	usedIDs, err := args.getUsedIDs(lockedEntries, owner)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// Add pending IDs to the used IDs to ensure we don't generate the same ID again
+	usedIDs = append(usedIDs, g.pendingIDs...)
+
+	usedIDs = normalizeUsedIDs(usedIDs, args.minID, args.maxID)
+
+	maxAttempts := min(maxIDGenerateIterations, args.maxID-args.minID+1)
+
+	for range maxAttempts {
+		id, pos, err := getIDCandidate(args.minID, args.maxID, usedIDs)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		available, err := args.isAvailableID(lockedEntries, id)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		if !available {
+			// Keep track of the id, keeping usedIDs, sorted, since we
+			// use the binary search to add elements to it.
+			usedIDs = slices.Insert(usedIDs, pos, id)
+
+			// If the GID is not available, try the next candidate
+			log.Debugf(context.Background(), "%s %d is already used", args.idType, id)
+			continue
+		}
+
+		g.pendingIDs = append(g.pendingIDs, id)
+		cleanup = func() {
+			idx := slices.Index(g.pendingIDs, id)
+			g.pendingIDs = append(g.pendingIDs[:idx], g.pendingIDs[idx+1:]...)
+		}
+		return id, cleanup, nil
+	}
+
+	return 0, nil, fmt.Errorf("failed to find a valid %s for after %d attempts",
+		args.idType, maxAttempts)
+}
+
+func getIDCandidate(minID, maxID uint32, usedIDs []uint32) (uint32, int, error) {
+	// Pick the preferred ID candidate, starting with minID.
+	preferredID := minID
+
+	if len(usedIDs) > 0 {
+		// If there are used IDs, we prefer the next ID above the highest used.
+		// Note that this may overflow, and so go back to 0, but the next check
+		// will adjust the value to the minID.
+		preferredID = usedIDs[len(usedIDs)-1] + 1
+
+		// Ensure that the preferred ID is not less than the minimum ID.
+		preferredID = max(preferredID, minID)
+	}
+
+	// Try IDs starting from the preferred ID up to the maximum ID.
+	for id := preferredID; id <= maxID; id++ {
+		if pos, found := slices.BinarySearch(usedIDs, id); !found {
+			return id, pos, nil
+		}
+
+		if id == math.MaxUint32 {
+			break // Avoid overflow
+		}
+	}
+
+	// Fallback: try IDs from the minimum ID up to the preferred ID.
+	for id := minID; id < preferredID && id <= maxID; id++ {
+		if pos, found := slices.BinarySearch(usedIDs, id); !found {
+			return id, pos, nil
+		}
+
```

Comment:
Ah, also as highlighted by coverage, it's impossible to reach this now, since `preferredID` can never been mayor than that, not sure if you want to keep it for consistency, but we have tests already using these values, so maybe we can just drop it.

Reply:
Mh diff seems scrambled, well I meant the

```go
		if id == math.MaxUint32 {
			break // Avoid overflow
		}
```

In the **second** loop